libraries/chain/fork_database.cpp

```diff
@@ -277,20 +277,22 @@ namespace eosio::chain {
          EOS_RETHROW_EXCEPTIONS( fork_database_exception, "serialized fork database is incompatible with configured protocol features" )
       }
 
+      auto prev_head = head_impl(include_root_t::no);
+
       auto inserted = index.insert(n);
       EOS_ASSERT(ignore_duplicate == ignore_duplicate_t::yes || inserted.second, fork_database_exception,
                  "duplicate block added: ${id}", ("id", n->id()));
+
+      return inserted.second && prev_head != head_impl(include_root_t::no);
```

Comment:
Actually we never need `prev_head`.

```suggestion
      return inserted.second && head_impl(include_root_t::no) == n;
```

Reply:
Or better yet, use iterator and avoid creating a `block_state_ptr`.

Reply:
I think creating the `shared_ptr` is very minimal overhead, not worth making the code more complicated.

Reply:
What do you think of the latest?

Reply:
You didn't remove the previous call to get the old head.